Feature: Offline buffering (aka Persistence mechanism)

svranesevic opened the pull request:
Source reorganized from flat to tree structure

MQTT messages are serialized from/to appropriate DTOs instead of internal models

Separation of responsibilities:
* ConnectivityService:
Responsible for sending/receiving messages from WolkAbout IoT Platform
* PublishService:
When data publish can be performed forwards readings to ConnectivityService
When data publish can not be performed, or fails, forwards readings to PersistService

utilities/FileSystemUtils:
Added file system utility functions

utilities/StringUti;s:
Added string utility functions

README.md:
Updated with new builder interface
Updated with new persistence functionality

CMakeLists.txt:
Added compiler warning flags

Commit:
Persistence feature implemented

Source reorganized from flat to tree structure

MQTT messages are serialized from/to appropriate DTO objects instead of internal models

Separation of responsibilities:
ConnectivityService - Responsible for sending/receiving messages from WolkAbout IoT Platform
PublishService - When data publish can be performed forwards readings to ConnectivityService
                 When data publish can not be performed, or fails, forwards readings to PersistService

utilities/FileSystemUtils:
Added file system utility functions

utilities/StringUti;s:
Added string utility functions

README.md:
Updated with new builder interface
Updated with new persistence functionality

Commit:
service/persist/json/JsonPersistService.cpp:
Persist only latest actuator status

milandinic reviewed (dismissed):
if found only commit comment to have misplaced semi-column.

Commit:
Improvements suggested by clang-tidy:
Removed moving (std::move) of trivially-copyable types
Deleted members are public
Use 'nullptr' instead of 'NULL'
Compare std::string using '==' operator
